pkg/serializer/internal/metrics/iterable_series_v3.go

```diff
@@ -0,0 +1,705 @@
+	var hash1, hash2 uint64 = uint64(prefixID), 0
+	for _, s := range tags {
+		hash1, hash2 = murmur3.SeedStringSum128(hash1, hash2, s)
+	}
+
+	key := hash1
+	if id, ok := db.tagsIndex[key]; ok {
+		return id
+	}
+
+	db.tagsBuffer = db.tagsBuffer[0:0]
+	if prefixID > 0 {
+		db.tagsBuffer = append(db.tagsBuffer, -prefixID)
+	}
+	db.appendTagsSlice(tags)
+	slices.Sort(db.tagsBuffer)
+	deltaEncode(db.tagsBuffer)
+
+	db.tagsLastID++
+	db.tagsIndex[key] = db.tagsLastID
+
+	db.txn.Sint64(columnDictTagsets, int64(len(db.tagsBuffer)))
+	for _, idx := range db.tagsBuffer {
+		db.txn.Sint64(columnDictTagsets, idx)
+	}
+
+	return db.tagsLastID
+}
+
+func (db *dictionaryBuilder) internTags(tags tagset.CompositeTags) int64 {
+	t1, t2 := tags.UnsafeGet()
+
+	if len(t1) == 0 && len(t2) == 0 {
+		return 0
+	} else if len(t1) == 0 {
+		return db.internTags1(0, t2)
+	} else if len(t2) == 0 {
+		return db.internTags1(0, t1)
+	}
+
+	db.stats.tagsSplit++
+	prefixID := db.internTags1(0, t1)
+	return db.internTags1(prefixID, t2)
+}
+
+func (db *dictionaryBuilder) internResources(res []metrics.Resource) int64 {
+	if len(res) == 0 {
+		return 0
+	}
+
+	var hash1, hash2 uint64
+	for _, r := range res {
+		hash1, hash2 = murmur3.SeedStringSum128(hash1, hash2, r.Type)
+		hash1, hash2 = murmur3.SeedStringSum128(hash1, hash2, r.Name)
+	}
+
+	key := hash1
+	if id, ok := db.resourcesIndex[key]; ok {
+		return id
+	}
+
+	db.resourcesLastID++
+	db.resourcesIndex[key] = db.resourcesLastID
+
+	db.txn.Int64(columnDictResourcesLen, int64(len(res)))
+
+	typeDelta := deltaEncoder{}
+	nameDelta := deltaEncoder{}
+
+	for _, res := range res {
+		db.txn.Sint64(columnDictResourceType,
+			typeDelta.encode(db.resourceInterner.intern(istr(res.Type))))
+		db.txn.Sint64(columnDictResourceName,
+			nameDelta.encode(db.resourceInterner.intern(istr(res.Name))))
+	}
+
+	return db.resourcesLastID
+}
+
+func (db *dictionaryBuilder) internOriginInfo(info originInfo) int64 {
+	return db.originInfoInterner.intern(info)
+}
+
+func (db *dictionaryBuilder) internSourceTypeName(stn string) int64 {
+	if stn == "" {
+		return 0
+	}
+	return db.sourceTypeNameInterner.intern(istr(stn))
+}
+
+func pointValueType(v float64) int64 {
+	if v == 0 {
+		return valueZero
+	}
+
+	// Integers in this range encode to 7 byte varints or less
+	const maxInt = 1<<48 - 1
+	const minInt = -1 << 48
+
+	i := int64(v)
+	if i >= minInt && i <= maxInt && float64(i) == v {
+		return valueSint64
+	}
+
+	if float64(float32(v)) == v {
+		return valueFloat32
+	}
+
+	return valueFloat64
+}
+
+func metricType(ty metrics.APIMetricType) int64 {
+	switch ty {
+	case metrics.APICountType:
+		return metricCount
+	case metrics.APIGaugeType:
+		return metricGauge
+	case metrics.APIRateType:
+		return metricRate
+	}
+	panic("unknown APIMetricType")
+}
+
+func varintLen(v int) int {
+	n, rem := bits.Div(0, uint(bits.Len(uint(v))), 7)
+	if rem > 0 {
+		n++
+	}
+	return int(n)
+}
```

Comment:
Not sure if we'd actually hit this, but this returns a zero length for an input value of zero. I believe that would still need one byte to encode.